core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.telemetry.impl.java11;
+
+import com.google.common.annotations.VisibleForTesting;
+import hudson.Extension;
+import jenkins.model.Jenkins;
+import jenkins.telemetry.Telemetry;
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Telemetry class to gather information about class loading issues when running on java 11.
+ * The use of this class is not restricted for external use because we may want to gather information from plugins
+ * directly by calling report... methods.
+ **/
+
+@Extension
+//@Restricted(NoExternalUse.class)
+public class MissingClassTelemetry extends Telemetry {
+    private static final Logger LOGGER = Logger.getLogger(MissingClassTelemetry.class.getName());
+
+    // Store 100 events today
+    private static MissingClassEvents events = new MissingClassEvents();
+
+    // When we begin to gather these data
+    private final static LocalDate START = LocalDate.of(2019, 4, 1);
+    // Gather for 2 years (who knows how long people will need to migrate to Java 11)
+    private final static LocalDate END = START.plusMonths(24);
+
+    // The types of exceptions which can be reported
+    private static final Set exceptionsToReportToJava11Telemetry =
```

Comment:
It is not Java11 telemetry, the same logic applies to all Java versions beyond Java9

Reply:
changed